### PR TITLE
gets started writing text on current injection

### PR DESCRIPTION
**Pull request by billbrod**
I got through the beginning part of this, opening PR so we can see it rendered.

### CODE REVIEW

File: docs/examples/01_current_injection.py [hunk 1]
```diff
@@ -1,502 +1,694 @@
...
+#
+# For this example, we'll use an un-regularized LBFGS solver. We'll discuss
+# regularization in a later tutorial.
+#
+# !!! info "Why LBFGS?"
+#
+#     LBFGS is a second-order optimizer, that is, it uses both the first
+#     derivative (the gradient) and the second derivative in order to solve the
+#     problem. This leads to a faster solution. Try other solvers to see how
+#     they behave!
```
Comment (ahwillia) [inline docs/examples/01_current_injection.py:483]: This is incorrect, LBFGS only uses gradients but it tries to approximate the Hessian by storing a recent history of gradients. It never explicitly computes a second derivative so it is called a "quasi-Newton method."

It would be good to add a link to the wikipedia page: https://en.wikipedia.org/wiki/Limited-memory_BFGS

For some of these models a true Newton's method should work very well. Have we played with this?

Fun aside: the "G" in LBFGS stands for Goldfarb and I guess he hangs around CCM. I randomly had lunch with him a couple weeks ago.

File: docs/examples/01_current_injection.py [hunk 2]
```diff
@@ -1,502 +1,694 @@
...
-# ### 3. The Poisson log-likelihood
-# The last component of the model is the poisson log-likelihood that quantifies how likely it is
-# to observe certain counts for a given firing rate time series.
-# The if $y(t)$ are the spike counts, the equation for the log-likelihood is
-# $$ \sum\_t \log P(y(t) | \lambda(t)) = \sum\_t  y(t) \log(\lambda(t)) - \lambda(t) - \log (y(t)!)\tag{2}$$
-# In nemos, the log-likelihood can be computed by calling the score method passing the predictors and the counts.
-# The method first compute the rate $\lambda(t)$ using (1) and then the likelihood using (2).
+#
+# Now that we've fit our data, we can retrieve the resulting parameters, the
+# analogues of the `weights` and `intercepts` variables used earlier in this
```
Comment (ahwillia) [inline docs/examples/01_current_injection.py:499]: This part seems to refer to the piece of the tutorial that got moved elsewhere?

File: docs/examples/01_current_injection.py [hunk 3]
```diff
@@ -1,502 +1,694 @@
...
+# `model.coef_` has shape `(n_neurons, n_features)`, while `model.intercept_`
+# has shape `(n_neurons)`, with those values coming from the model matrix we
+# passed during fitting.
+#
+# It's nice to get the parameters above, but we can't tell how well our model
+# is doing by looking at them. So how should we evaluate our model?
+#
+# First, we can use the model to predict the firing rates and compare that to
+# our observed firing rate, to see which of the phenomena we described in our
+# pynapple analysis our model is able to capture. By calling `predict()` we can
```
Comment (ahwillia) [inline docs/examples/01_current_injection.py:524]: I suggest that we change "observed firing rate" -> "smoothed spike train" (or something like this) everywhere.

Also I think this sentence can be shortened a bit

```suggestion
# First, we can use the model to predict the firing rates and compare that to
# the smoothed spike train. By calling `predict()` we can
```

File: docs/examples/01_current_injection.py [hunk 4]
```diff
@@ -1,502 +1,694 @@
...
+# let's reintroduce the time axis by defining a TsdFrame.
 
-# let's reintroduce the time axis by defining a TsdFrame
-# convert first to numpy array otherwise to make it pynapple compatible
-predicted_fr = nap.TsdFrame(t=count.t, d=np.asarray(predicted_fr))
+# we must convert the firing rate to a numpy array (from jax.numpy) to make it
+# pynapple compatible
+predicted_fr = nap.TsdFrame(t=binned_current.t, d=np.asarray(predicted_fr))
+# and let's smooth the firing rate the same way that we smoothed the observed
+# firing rate
```
Comment (ahwillia) [inline docs/examples/01_current_injection.py:538]: "observed firing rate" -> "smoothed spike train"

File: docs/examples/01_current_injection.py [hunk 5]
```diff
@@ -1,502 +1,694 @@
...
 smooth_predicted_fr = predicted_fr.smooth(50, 1000)
 
+# and plot!
+utils.plotting.current_injection_plot(current, spikes, firing_rate,
+                                      # plot the predicted firing rate that has
+                                      # been smoothed the same way as the observed
```
Comment (ahwillia) [inline docs/examples/01_current_injection.py:544]: "observed firing rate" -> "smoothed spike train"

File: docs/examples/01_current_injection.py [hunk 6]
```diff
@@ -1,502 +1,694 @@
...
+# each subplot!
+#
+# - Predicted firing rate increases as injected current goes up -- Success!
+#
+# - The amplitude of the predicted firing rate only matches the observed
+#   amplitude in the third interval: it's too high in the first and too low in
+#   the second -- Failure!
+#
+# - Our predicted firing rate has the periodicity we see in the observed firing
+#   rate -- Success!
```
Comment (ahwillia) [inline docs/examples/01_current_injection.py:560]: "observed firing rate" -> "smoothed spike train"

File: docs/examples/01_current_injection.py [hunk 7]
```diff
@@ -1,502 +1,694 @@
...
-# # then add self-excitatin
+# %%
+#
+# In addition to making that mismatch discussed earlier a little more obvious,
+# this tuning curve comparison also highlights that this model thinks the
+# firing rate will continue to grow as the injected current increases, which is
+# not reflected in the data.
+#
+# Viewing this plot also makes it clear that the model's tuning curve is an
+# exponential. We already knew that! That's what it means to be a LNP model of
```
Comment (ahwillia) [inline docs/examples/01_current_injection.py:600]: change to "approximately an exponential" or something similar.

File: docs/examples/01_current_injection.py [hunk 8]
```diff
@@ -1,502 +1,694 @@
...
+
+spikes = jax.random.poisson(jax.random.PRNGKey(0), predicted_fr.values)
+
+# %%
+#
+# Note that this is not actually that informative and, in general, it is
+# recommended that you focus on firing rates when interpreting your model. In
+# particular, if your GLM includes auto-regressive inputs (e.g., neurons are
+# connected to themselves or each other), simulate can behave poorly, see
+# **XXX** for details.
```
Comment (ahwillia) [inline docs/examples/01_current_injection.py:619]: Maybe this could just be put under "exercises"?

File: docs/examples/01_current_injection.py [hunk 9]
```diff
@@ -1,502 +1,694 @@
...
+#     Under the hood, nemos is minimizing the negative log-likelihood, as is
+#     typical in many optimization contexts. `score` returns the real
+#     log-likelihood, however, and thus higher is better.
+#
+# Because it's un-normalized, however, the log-likelihood should not be
+# compared across datasets (because e.g., it won't account for difference in
+# noise levels). We provide the ability to compute the pseudo-$R^2$ for this
+# purpose:
+
+model.score(predictor, count, score_type='pseudo-r2-Cohen')
```
Comment (ahwillia) [inline docs/examples/01_current_injection.py]: I think this part might want to be placed into its own section called "Evaluating the model on train and test sets"

**Review by ahwillia — COMMENTED** [review]
Looks really nice Billy thanks for putting in so much time to make this readable and approachable 👍

File: docs/examples/01_current_injection.py [hunk 10]
```diff
@@ -1,502 +1,694 @@
...
+#     as the literal truth (to see why, pass a firing rate through a Poisson
+#     process to generate spikes and then smooth the output to approximate the
+#     generating firing rate). A model should not be expected to match this
+#     approximate firing rate exactly, but visualizing the two firing rates
+#     together can help you reason about which phenomena in your data the model
+#     is able to adequately capture, and which it is missing.
+#
+#     For more information, see section 1.2 of [*Theoretical
+#     Neuroscience*](https://boulderschool.yale.edu/sites/default/files/files/DayanAbbott.pdf),
+#     by Dayan and Abbott.
```
[the remark this answers is not in the document]
Comment (billbrod) [reply docs/examples/01_current_injection.py:253]: I was just trying to stress that comparing the predicted and smoothed firing rates is a heuristic, and you shouldn't worry too much if they don't match exactly -- you don't think that's necessary to say?

File: docs/examples/01_current_injection.py [hunk 11]
```diff
@@ -1,502 +1,694 @@
...
+#   beliefs about the parameters, such as sparsity. Regularization becomes more
+#   important as the number of input features, and thus model parameters,
+#   grows. They can be found within `nemos.regularizer`.
+#
+# !!! warning
+#
+#     With a convex problem like the GLM, in theory it does not matter which
+#     solver algorithm you use. In practice, due to numerical issues, it might.
+#     Thus, it's worth trying a couple to see how their solutions compare.
+#     (Different regularization schemes will always give different results.)
```
[the remark this answers is not in the document]
Comment (billbrod) [reply docs/examples/01_current_injection.py:461]: hmm. I wanted to highlight this in case people were confused about why we set the optimizer at all -- when we described the model as convex, one of the selling points was that we'd always find the best solution, so why does it matter.

I think what you're describing sounds like a good idea, but too much to get done before this workshop. I copied your commnets into https://github.com/flatironinstitute/nemos/issues/90 and we'll address it after the workshop.

For this workshop, do you think this caveat is okay?

File: docs/examples/01_current_injection.py (hunk 1 above)
Comment (billbrod) [reply docs/examples/01_current_injection.py:480]: How's the change I made?

File: docs/examples/01_current_injection.py (hunk 2 above)
Comment (billbrod) [reply docs/examples/01_current_injection.py:496]: whoops ,yes, good catch

File: docs/examples/01_current_injection.py (hunk 8 above)
Comment (billbrod) [reply docs/examples/01_current_injection.py:616]: why so?

File: docs/examples/01_current_injection.py [hunk 12]
```diff
@@ -1,502 +1,694 @@
...
+#
+# Similar to `current`, this object originally contains data from the entire
+# experiment. To get only the data we need, we again use
+# `restrict(noise_interval)`:
 
-noise_interval = noise_interval.loc[[0]]
 spikes = spikes.restrict(noise_interval)
-current = current.restrict(noise_interval)
+print(spikes)
+spikes[1]
```
[the remark this answers is not in the document]
Comment (billbrod) [reply docs/examples/01_current_injection.py:200]: I could index into it (`spikes[1][:50]`) but that's more complicated. I'm unsure if there's a global variable to determine box height.

@gviejo there is some way to determine when pynapple switches to "abbreviated", right? does that work with `Ts` objects?

File: docs/examples/01_current_injection.py (hunk 9 above)
Comment (billbrod) [reply docs/examples/01_current_injection.py:649]: Is this section still part of the main text or placed under "Further exercises" with the "Train and test sets" bit?

File: docs/examples/01_current_injection.py [hunk 13]
```diff
@@ -1,502 +1,694 @@
...
 # %%
-# As you can see there are 3 rows. Each contains the start and end (in seconds) of an epoch.
-# To select only one epoch from an IntervalSet, use 2 square brackets.
+#
+# As described above, we will be examining "Noise 1". We can see it contains
+# three rows, each defining a separate sweep. We'll just grab the first sweep
+# (shown in blue in the pictures above) and ignore the other two (shown in
+# gray).
+#
+# To select only one epoch from an IntervalSet, use 2 square brackets:
```
[the remark this answers is not in the document]
Comment (gviejo) [reply docs/examples/01_current_injection.py:145]: IntervalSet are still pandas. Changing them to numpy array is in the roadmap.

File: docs/examples/01_current_injection.py [hunk 14]
```diff
@@ -1,502 +1,694 @@
...
-# TsGroup is a dictionnary. You can look at the spike times of neuron by indexing it.
-print(spikes[1])
+#
+# Typically, this is used to hold onto the spike times for a population of
+# neurons. In this experiment, we only have recordings from a single neuron, so
+# there's only one row.
+#
+# We can index into the `TsGroup` to see the timestamps for this neuron's
+# spikes:
+spikes[1]
```
[the remark this answers is not in the document]
Comment (gviejo) [reply docs/examples/01_current_injection.py:190]: No big reason. I think it comes from the original NWB file we transformed.
